README
GeeXboX uShare - Introduction
=============================

GeeXboX uShare is a UPnP (TM) A/V Media Server. It implements the server
component that provides UPnP media devices with information on available
multimedia files. uShare uses the built-in http server of libupnp to
stream the files to clients.

GeeXboX uShare is able to provide access to both images, videos, music
or playlists files (see below for a complete file format support list).
It does not act as an UPnP Media Adaptor and thus, can't transcode
streams to fit the client requirements.

uShare is written in C for the GeeXboX project (see http://www.geexbox.org/).
It is designed to provide access to multimedia contents to GeeXboX but can of
course be used by any other UPnP client device.
It should compile and run on any modern POSIX compatible system such as Linux.

GeeXboX uShare is free software - it is licensed under the terms of the GNU
General Public License (GPL).

Copyright and License
=====================

GeeXboX uShare is copyright (C) 2005-2007 Benjamin Zores.

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your option)
any later version.

This program is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc., 51
Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

Note that uShare links with libupnp, which is licensed under the
terms of a modified BSD license (i.e. the original BSD license without the
advertising clause). This license is compatible with the GNU GPL.

Homepage
========

Web site and file area for uShare is hosted on GeeXboX server :

   http://ushare.geexbox.org/

The latest version of uShare should always be available on this site.

Requirements
============

The following programs are required to build uShare:

 * GNU C Compiler (gcc), 2.95 or later.

   The GNU C Compiler is part of the GNU Compiler Collection which can be
   downloaded from http://gcc.gnu.org/.

The following UPnP library is required to build and run uShare :

 * Linux SDK for UPnP Devices (libupnp), 1.4.2 or later

   The libupnp library is used to communicate using the UPnP protocol.
   libupnp can be downloaded from http://pupnp.sourceforge.net/.

The following DLNA library is required for proper DLNA support :

 * libdlna, 0.2.1 or later

   The libdlna library is used to provides DLNA profiles informations.
   libdlna can be downloaded from http://libdlna.geexbox.org/.

Building and Installation
=========================

Compile uShare by running configure and then make. This should
produce an executable ushare in the src subdirectory, which can be
used right away. No extra files need to be installed.

You can pass the CFLAGS you want to configure including -DDEBUG in order
to activate support for debug messages in uShare.

Example :

CFLAGS="-Os" ./configure --prefix=/usr
make

You can enable DLNA support by doing a:
./configure --enable-dlna

If you want to install uShare on your system, run :

make install-strip

This will copy the executable and manual page into their appropriate
directories (/usr/bin and /usr/man/man1 in this example).

For more information regarding configure and make, see the INSTALL document.

Usage
=====

uShare runs from the console only. It supports the usual --help option
which displays usage and option information.

Usage: ushare [-n name] [-i interface] [-c directory] [[-c directory] ...]
Options:
 -n, --name=NAME        Set UPnP Friendly Name (default is 'uShare')
 -i, --interface=IFACE  Use IFACE Network Interface (default is 'eth0')
 -f, --cfg=FILE         Config file to be used
 -p, --port=PORT        Forces the HTTP server to run on PORT
 -q, --telnet-port=PORT Forces the TELNET server to run on PORT
 -c, --content=DIR      Share the content of DIR directory.
 -w, --no-web           Disable the control web page (enabled by default)
 -t, --no-telnet        Disable the TELNET control (enabled by default)
 -o, --override-iconv-err       If iconv fails parsing name, still add to media contents (hoping the renderer can handle it)
 -v, --verbose          Set verbose display
 -x, --xbox             Use XboX 360 compliant profile
 -d, --dlna             Use DLNA compliant profile (PlayStation3 needs this)
 -D, --daemon           Run as a daemon.
 -V, --version          Display the version of uShare and exit
 -h, --help             Display this help

uShare gets its configuration from the /etc/ushare.conf file.
You can force configuration options through command line.

uShare expects one or several directory argument (-c argument),
specifying where multimedia files are stored. You should probably also use
the -i option to specify which interface uShare should listen on.

   ushare -c /shares
   ushare -c /shares1 --content=/shares2

You can also perform remote control of uShare UPnP Media Server through its
web interface. This let you define new content locations at runtime or
update the currently shared one in case the filesystem has changed.
Just go to :

   http://ip_address:port/web/ushare.html

See the manual page for more details :

   man ushare

Supported File Formats List
===========================

- Video files : asf, avi, dv, divx, wmv, mjpg, mjpeg, mpeg, mpg, mpe,
                mp2p, vob, mp2t, m1v, m2v, m4v, m4p, mp4ps, ts, ogm, mkv,
                rmvb, mov, qt

- Audio files : aac, ac3, aif, aiff, at3p, au, snd, dts, rmi, mp1, mp2, mp3,
                mp4, mpa, ogg, wav, pcm, lpcm, l16, wma, mka, ra, rm, ram

- Images files : bmp, ico, gif, jpeg, jpg, jpe, pcd, png, pnm, ppm,
                 qti, qtf, qtif, tif, tiff

- Playlist files : pls, m3u, asx

If you want uShare to support more file formats, simply add its properties
in the src/mime.c table. Do not forget to send a patch to update uShare.

Feedback
========

Please send bug reports, suggestions, ideas, comments or patches to :

   ushare@geexbox.org

Known bugs and limitations
==========================

If you need to listen on more than one interface, you will have to start
multiple instances of the media server.

uShare keeps some information on files in memory.
If your multimedia collection is huge, this might be a problem.

Thanks
======

Many thanks to Oskar Liljeblad for its original work on GMediaServer
(which is much more functionnal in terms of UPnP A/V features that
uShare will ever be).

References
==========

Note that this list of references is not complete.

 * UPnP(TM) Standards (http://www.upnp.org/standardizeddcps/default.asp)

Trademarks
==========

UPnP(TM) is a trademark of the UPnP(TM) Implementers Corporation.

-


NEWS
2007-12-04: GeeXboX uShare 1.1 released.
 Much more complete DLNA support through external libdlna.
 Telnet Control interface.
 Support for FLAC and HDMOV files.
 FreeBSD 64bit and MacOSX build fix.
 Newly written configure script.

2007-07-05 : GeeXboX uShare 1.0 released.
 Fixed XboX 360 file discovery (it finally should work).
 Support for Sony PlayStation 3.
 DLNA compliance.

2007-02-26 : GeeXboX uShare 0.9.10 released.
 Fixed http 404 error which prevented ushare 0.9.9 release to work.

2007-02-25 : GeeXboX uShare 0.9.9 released.
 Added support for FreeBSD.
 Use a RedBlack balanced tree searching and sorting algorithm
  instead of the recursive one. It avoids recursive calls, takes
  less memory (perfect for embedded systems) and executes in more
  "constant time" than the logarithmic style.
  Original patch by Phil Chandler.	
 Massive memory optimization (uShare now takes 75% less RAM ;-)
  only not using fixed size URL allocation buffer (a 16000+ files
  collection now only takes 5MB where it used to take 19MB).
 Improved search parsing and matching mechanism (patch by Mostafa
  Hosseini). XboX360 support is now complete.
  Please note that to stream video to the XboX, the console should
  be running dashboard version 2.0.4552.0 and only unprotected
  .wmv files are supported.
 Fixed support for 2GB+ files (see http://sourceforge.net/tracker/index.php?func=detail&aid=1634927&group_id=151880&atid=782410)
 Fixed support for XboX 360 (thanks to Mostafa Hosseini).
 Added basic support for 'Search' action.
 Fixed bug with localization, use now --localedir configure's
  option to set location of .mo files.
 Make libupnp >= 1.4.2 a requirement.

2006-12-10 : GeeXboX uShare 0.9.8 released.
 Added German translation.
 Convert special xml characters to xml (like & => &amp;).
 Added iso and bin mime types.
 Better filtering and browse capabilities: now uShare is conform to
  ContentDirectory:1 Service Template Standard.
 Updated init script to use lsb functions.

2006-03-12 : GeeXboX uShare 0.9.7 released.
 Added new Microsoft Registrar service which fakes the authorization
  required to be able to stream to MS compliant Media Renderers.
 Updated AVI file MIME type for better compliance with UPnP devices.
 Better UPnP CDS compliance when request count and index are 0.
 Removed the nasty characters from UDN naming for compatibility issues.
 Force usage of libupnp >= 1.3.1
  (previous versions of the library are buggy and should not be used).
 Support for XboX 360, D-Link DSM-320 and Roku SoundBridge M2000
  Network Music Player devices.

2006-02-19 : GeeXboX uShare 0.9.6 released.
 Added Fedora RPM build and startup init scripts.
 Support for for large files (2GB+ files).
 Can now specify on which port the HTTP server has to run.
 Fixed a potential buffer overrun when building DIDL messages.

2006-01-31 : uShare has been included in the Fedora Project.
 See http://fedoraproject.org/extras/4/i386/repodata/repoview/ushare-0-0.9.5-4.fc4.html

2005-12-19 : GeeXboX uShare 0.9.5 released.
 Added UPnP presentation URL support (you can now update or add/remove
 contents through a web interface).
 Added support for subtitle file formats.
 Added support for DVD specific file formats.

2005-12-07 : GeeXboX uShare 0.9.4 released.
 Fixed some memleaks.
 Avoid having empty shared directory name.
 Fixed handling of non-absolute content directories paths.
 Added logging with syslog support, in daemon mode.

2005-11-20: GeeXboX uShare 0.9.3 released.
 Support for multiple directories to be shared.
 Added start/stop script.
 Added verbose mode.
 Read configuration from /etc/ushare.conf configuration file.
 Support for daemon mode.
 Better MIME types handling.
 Support for new multimedia files extensions.
 Rewrite of some string buffer and integer manipulation tools.                

2005-11-13: GeeXboX uShare 0.9.2 released.
 Improved UPnP compliance with Browse RequestedCount flag.
 Fixed some UPnP Object ID sent in Browse messages (terrible mistake).

2005-11-10: GeeXboX uShare 0.9.1 released.
 uShare ready for internationalization.
 Added French localization.
 Support of iconv, for UTF-8 filenames conversions.
 Added Debian package building scripts.

2005-10-25: GeeXboX uShare 0.9.0 released.
  First public release.
  Tested with :
    - Intel Digital Media Adaptor
    - Intel UPnP AV Media Controller


src/mime.c
/*
 * mime.c : GeeXboX uShare media file MIME-type association.
 * Originally developped for the GeeXboX project.
 * Ref : http://freedesktop.org/wiki/Standards_2fshared_2dmime_2dinfo_2dspec
 * Copyright (C) 2005-2007 Benjamin Zores <ben@geexbox.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Library General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include <stdlib.h>
#include <string.h>

#include "mime.h"
#include "ushare.h"

#define UPNP_VIDEO "object.item.videoItem"
#define UPNP_AUDIO "object.item.audioItem.musicTrack"
#define UPNP_PHOTO "object.item.imageItem.photo"
#define UPNP_PLAYLIST "object.item.playlistItem"
#define UPNP_TEXT "object.item.textItem"

const struct mime_type_t MIME_Type_List[] = {
  /* Video files */
  { "asf",   UPNP_VIDEO, "http-get:*:video/x-ms-asf:"},
  { "avc",   UPNP_VIDEO, "http-get:*:video/avi:"},
  { "avi",   UPNP_VIDEO, "http-get:*:video/avi:"},
  { "dv",    UPNP_VIDEO, "http-get:*:video/x-dv:"},
  { "divx",  UPNP_VIDEO, "http-get:*:video/avi:"},
  { "wmv",   UPNP_VIDEO, "http-get:*:video/x-ms-wmv:"},
  { "mjpg",  UPNP_VIDEO, "http-get:*:video/x-motion-jpeg:"},
  { "mjpeg", UPNP_VIDEO, "http-get:*:video/x-motion-jpeg:"},
  { "mpeg",  UPNP_VIDEO, "http-get:*:video/mpeg:"},
  { "mpg",   UPNP_VIDEO, "http-get:*:video/mpeg:"},
  { "mpe",   UPNP_VIDEO, "http-get:*:video/mpeg:"},
  { "mp2p",  UPNP_VIDEO, "http-get:*:video/mp2p:"},
  { "vob",   UPNP_VIDEO, "http-get:*:video/mp2p:"},
  { "mp2t",  UPNP_VIDEO, "http-get:*:video/mp2t:"},
  { "m1v",   UPNP_VIDEO, "http-get:*:video/mpeg:"},
  { "m2v",   UPNP_VIDEO, "http-get:*:video/mpeg2:"},
  { "mpg2",  UPNP_VIDEO, "http-get:*:video/mpeg2:"},
  { "mpeg2", UPNP_VIDEO, "http-get:*:video/mpeg2:"},
  { "m4v",   UPNP_VIDEO, "http-get:*:video/mp4:"},
  { "m4p",   UPNP_VIDEO, "http-get:*:video/mp4:"},
  { "mp4",   UPNP_VIDEO, "http-get:*:video/mp4:"},
  { "mp4ps", UPNP_VIDEO, "http-get:*:video/x-nerodigital-ps:"},
  { "ts",    UPNP_VIDEO, "http-get:*:video/mpeg2:"},
  { "ogm",   UPNP_VIDEO, "http-get:*:video/mpeg:"},
  { "mkv",   UPNP_VIDEO, "http-get:*:video/mpeg:"},
  { "rmvb",  UPNP_VIDEO, "http-get:*:video/mpeg:"},
  { "mov",   UPNP_VIDEO, "http-get:*:video/quicktime:"},
  { "hdmov", UPNP_VIDEO, "http-get:*:video/quicktime:"},
  { "qt",    UPNP_VIDEO, "http-get:*:video/quicktime:"},
  { "bin",   UPNP_VIDEO, "http-get:*:video/mpeg2:"},
  { "iso",   UPNP_VIDEO, "http-get:*:video/mpeg2:"},

  /* Audio files */
  { "3gp",  UPNP_AUDIO, "http-get:*:audio/3gpp:"},
  { "aac",  UPNP_AUDIO, "http-get:*:audio/x-aac:"},
  { "ac3",  UPNP_AUDIO, "http-get:*:audio/x-ac3:"},
  { "aif",  UPNP_AUDIO, "http-get:*:audio/aiff:"},
  { "aiff", UPNP_AUDIO, "http-get:*:audio/aiff:"},
  { "at3p", UPNP_AUDIO, "http-get:*:audio/x-atrac3:"},
  { "au",   UPNP_AUDIO, "http-get:*:audio/basic:"},
  { "snd",  UPNP_AUDIO, "http-get:*:audio/basic:"},
  { "dts",  UPNP_AUDIO, "http-get:*:audio/x-dts:"},
  { "rmi",  UPNP_AUDIO, "http-get:*:audio/midi:"},
  { "mid",  UPNP_AUDIO, "http-get:*:audio/midi:"},
  { "mp1",  UPNP_AUDIO, "http-get:*:audio/mp1:"},
  { "mp2",  UPNP_AUDIO, "http-get:*:audio/mp2:"},
  { "mp3",  UPNP_AUDIO, "http-get:*:audio/mpeg:"},
  { "m4a",  UPNP_AUDIO, "http-get:*:audio/mp4:"},
  { "ogg",  UPNP_AUDIO, "http-get:*:audio/x-ogg:"},
  { "wav",  UPNP_AUDIO, "http-get:*:audio/wav:"},
  { "pcm",  UPNP_AUDIO, "http-get:*:audio/l16:"},
  { "lpcm", UPNP_AUDIO, "http-get:*:audio/l16:"},
  { "l16",  UPNP_AUDIO, "http-get:*:audio/l16:"},
  { "wma",  UPNP_AUDIO, "http-get:*:audio/x-ms-wma:"},
  { "mka",  UPNP_AUDIO, "http-get:*:audio/mpeg:"},
  { "ra",   UPNP_AUDIO, "http-get:*:audio/x-pn-realaudio:"},
  { "rm",   UPNP_AUDIO, "http-get:*:audio/x-pn-realaudio:"},
  { "ram",  UPNP_AUDIO, "http-get:*:audio/x-pn-realaudio:"},
  { "flac", UPNP_AUDIO, "http-get:*:audio/x-flac:"},

  /* Images files */
  { "bmp",  UPNP_PHOTO, "http-get:*:image/bmp:"},
  { "ico",  UPNP_PHOTO, "http-get:*:image/x-icon:"},
  { "gif",  UPNP_PHOTO, "http-get:*:image/gif:"},
  { "jpeg", UPNP_PHOTO, "http-get:*:image/jpeg:"},
  { "jpg",  UPNP_PHOTO, "http-get:*:image/jpeg:"},
  { "jpe",  UPNP_PHOTO, "http-get:*:image/jpeg:"},
  { "pcd",  UPNP_PHOTO, "http-get:*:image/x-ms-bmp:"},
  { "png",  UPNP_PHOTO, "http-get:*:image/png:"},
  { "pnm",  UPNP_PHOTO, "http-get:*:image/x-portable-anymap:"},
  { "ppm",  UPNP_PHOTO, "http-get:*:image/x-portable-pixmap:"},
  { "qti",  UPNP_PHOTO, "http-get:*:image/x-quicktime:"},
  { "qtf",  UPNP_PHOTO, "http-get:*:image/x-quicktime:"},
  { "qtif", UPNP_PHOTO, "http-get:*:image/x-quicktime:"},
  { "tif",  UPNP_PHOTO, "http-get:*:image/tiff:"},
  { "tiff", UPNP_PHOTO, "http-get:*:image/tiff:"},

  /* Playlist files */
  { "pls", UPNP_PLAYLIST, "http-get:*:audio/x-scpls:"},
  { "m3u", UPNP_PLAYLIST, "http-get:*:audio/mpegurl:"},
  { "asx", UPNP_PLAYLIST, "http-get:*:video/x-ms-asf:"},

  /* Subtitle Text files */
  { "srt", UPNP_TEXT, "http-get:*:text/srt:"}, /* SubRip */
  { "ssa", UPNP_TEXT, "http-get:*:text/ssa:"}, /* SubStation Alpha */
  { "stl", UPNP_TEXT, "http-get:*:text/srt:"}, /* Spruce */
  { "psb", UPNP_TEXT, "http-get:*:text/psb:"}, /* PowerDivX */
  { "pjs", UPNP_TEXT, "http-get:*:text/pjs:"}, /* Phoenix Japanim */
  { "sub", UPNP_TEXT, "http-get:*:text/sub:"}, /* MicroDVD */
  { "idx", UPNP_TEXT, "http-get:*:text/idx:"}, /* VOBsub */
  { "dks", UPNP_TEXT, "http-get:*:text/dks:"}, /* DKS */
  { "scr", UPNP_TEXT, "http-get:*:text/scr:"}, /* MACsub */
  { "tts", UPNP_TEXT, "http-get:*:text/tts:"}, /* TurboTitler */
  { "vsf", UPNP_TEXT, "http-get:*:text/vsf:"}, /* ViPlay */
  { "zeg", UPNP_TEXT, "http-get:*:text/zeg:"}, /* ZeroG */
  { "mpl", UPNP_TEXT, "http-get:*:text/mpl:"}, /* MPL */

  /* Miscellaneous text files */
  { "bup", UPNP_TEXT, "http-get:*:text/bup:"}, /* DVD backup */
  { "ifo", UPNP_TEXT, "http-get:*:text/ifo:"}, /* DVD information */

  { NULL, NULL, NULL}
};

char *mime_get_protocol (struct mime_type_t *mime)
{
  char protocol[512];
  
  if (!mime)
    return NULL;

  sprintf (protocol, mime->mime_protocol);
  strcat (protocol, "*");
  return strdup (protocol);
}
